### PR TITLE
ExplicitTypeInterfaceRule rule

### PR DESCRIPTION
**Pull request by kimdv**
Hello! 

I don't want to start a war here, so I made this an opt in as default!

To explain myself briefly; I added this rule because I think it makes the code more readable and helps keep an overview of the specific properties. 

As stated before - the rule is made an OPT IN, so people themselves can decide whether to follow the rule or not. 

I have yet to figure out how the test works with the `nonTriggeringExamples`, `triggeringExamples` and `↓`. 

I hope that there is a nice person, who will help me out here, so we can finish the rule and give the people who like it, an opportunity to start using it. :)

Remember to keep it civil. 👍

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/TypeInterfaceRule.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    public func validate(file: File, kind: SwiftDeclarationKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        guard kind == .varInstance else {
+            return []
+        }
+
+        // Check if the property have a type
+        if dictionary.typeName != nil {
+                return []
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/TypeInterfaceRule.swift]: too much indentation

**Comment by marcelofabri** [comment]
What do you think about renaming it to `ExplicitTypeInterfaceRule` or something like that? Mainly because in the future we may want to have a rule to enforce the opposite.

Also, is there a reason why other kind of variables aren't validated? IMO at least class/static variables should be validate too.

**Comment by kimdv** [comment]
@marcelofabri I have changed the name to `ExplicitTypeInterfaceRule ` and added more var types. 

But I still don't know how the tests work, can you please help me out ? :)

**Comment by marcelofabri** [comment]
> But I still don't know how the tests work, can you please help me out ? :)

I think you've figured it out, since the tests are now passing 😅 

Can you rebase this to fix the conflicts?

**Comment by kimdv** [comment]
hmm, it seems it fails the checks?

Is the tests made wrong? :)


I also resolved the conflicts, so it can be merged with the master.

**Comment by marcelofabri** [comment]
@kimdv I haven't checked it locally, but my guess is that you need to wrap `class var` and `static var` declarations instead of a class (or a type for `static`).

You can run the tests on Xcode to debug it further.

**Comment by kimdv** [comment]
@marcelofabri I see that all the tests fails?

```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
let ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
static var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
class var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
let ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
static var ↓myVar = 0
```
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:256: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : XCTAssertEqual failed: ("0") is not equal to ("1") - 
/Users/travis/build/realm/SwiftLint/Tests/SwiftLintFrameworkTests/TestHelpers.swift:252: error: -[SwiftLintFrameworkTests.RulesTests testExplicitTypeInterfaceRule] : failed - triggeringExample did not violate at expected location: 
```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class var ↓myVar = 0
```
```

**Comment by marcelofabri** [comment]
@kimdv probably for the same reason: if you don't wrap those into a type, they'll be `.varGlobal`, which the rule doesn't handle.

**Comment by marcelofabri** [comment]
Can you pleas explain or tell me where I can read about the `↓` and how it works? :)

This marker is used to indicate where the violation should be reported (instead of only validating that a violation is reported).

For some reason the tests were falling on Travis even tough they succeeded on my machine. I've pushed d11b24d which should fix the only failure that happened here.

File: Source/SwiftLintFramework/Models/MasterRuleList.swift [hunk 2]
```diff
@@ -143,6 +143,7 @@ public let masterRuleList = RuleList(rules:
     TrailingSemicolonRule.self,
     TrailingWhitespaceRule.self,
     TypeBodyLengthRule.self,
+    ExplicitTypeInterfaceRule.self,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/MasterRuleList.swift]: this should be kept ordered

File: Tests/SwiftLintFrameworkTests/RulesTests.swift [hunk 3]
```diff
@@ -419,7 +423,8 @@ extension RulesTests {
             ("testVerticalParameterAlignment", testVerticalParameterAlignment),
             ("testVoidReturn", testVoidReturn),
             ("testSuperCall", testSuperCall),
-            ("testWeakDelegate", testWeakDelegate)
+            ("testWeakDelegate", testWeakDelegate),
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: this should be kept ordered

File: Tests/SwiftLintFrameworkTests/RulesTests.swift [hunk 4]
```diff
@@ -346,6 +346,10 @@ class RulesTests: XCTestCase {
         verifyRule(WeakDelegateRule.description)
     }
 
+    func testExplicitTypeInterfaceRule() {
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: this should be kept ordered

**Comment by kimdv** [comment]
@marcelofabri The test and rules should be ordered now. 

Lets see if the test passes now 😃

**Comment by kimdv** [comment]
hmm, still one test failing...??

**Comment by marcelofabri** [comment]
@kimdv Actually the tests were failing when using SwiftLint with Swift 2.3. I took the liberty to fix them and do some small refactorings on your branch, if that's ok.

@jpsim Since it's a new rule, would you like to review this as well?

**Comment by marcelofabri** [comment]
Actually from some reason I can't push to your fork from this computer, so I got tired of fighting against git and opened #1356

**Comment by marcelofabri** [comment]
Closing this in favor of #1356

**Pull request closed without merging**